Tests/Client/Unit Tests/AtomicTests.swift

```diff
@@ -0,0 +1,234 @@
+    }
+
+    func test_Atomic_keyPathHelpers() {
+        // Setup
+        struct Helper: Equatable {
+            var elements: [Int] = []
+        }
+        let value = Atomic(Helper())
+
+        // Actions
+        value.update(\.elements, to: [0])
+        XCTAssertEqual(value.get(), Helper(elements: [0]))
+
+        value.elements = [1]
+        XCTAssertEqual(value.get(), Helper(elements: [1]))
+
+        value.append(to: \.elements, 2)
+        XCTAssertEqual(value.get(), Helper(elements: [1, 2]))
+    }
+
+    func test_Atomic_dictionaryHelpers() {
+        let value = Atomic(["Luke": 1])
+        XCTAssertEqual(value["Luke"], 1)
+    }
+
+    func test_Atomic_whenCalledRecursively_worksCorrectly() {
+        // Setup
+        let value = Atomic("Luke")
+
+        // Action
+        value.update { (oldValue) -> String? in
+            // The current value should be same as `oldValue`
+            XCTAssertEqual(value.get(), oldValue)
+
+            value.set("Skywalker")
+            // The current value should be updated
+            XCTAssertEqual(value.get(), "Skywalker")
+
+            value.update { oldValue in
+                // The current value should be same as `oldValue`
+                XCTAssertEqual(value.get(), oldValue)
+                return "Vader!"
+            }
+            // The current value should be updated
+            XCTAssertEqual(value.get(), "Vader!")
+
+            // Finally, the return value from `update` closure should override previous changes
+            return "Leia"
+        }
+
+        XCTAssertEqual(value.get(), "Leia")
+    }
+}
+
+// MARK: - Stress tests
+
+extension AtomicTests {
+
+    var numberOfStressTestCycles: Int { 50 }
+
+    // Uncomment the test below to run the full stress test suite.
+    // Don't forget to increase `numberOfStressTestCycles` to some more interesting value.
+
+//    func test_Atomic_underHeavyLoad_behavesCorrently() {
+//        for _ in 0..<100 {
+//            test_Atomic_usedAsCounter_behavesCorrectly()
+//            test_Atomic_usedWithCollection_behavesCorrectly()
+//            test_Atomic_whenSetAndGetCalledSimultaneously_behavesCorrectly()
+//            test_Atomic_whenCalledFromMainThred_behavesCorrectly()
+//        }
+//    }
+
+    func test_Atomic_usedAsCounter_behavesCorrectly() {
+        var value: Atomic<Int>! = .init(0)
+
+        // Count up to numberOfCycles
+        for _ in 0..<numberOfStressTestCycles {
+            DispatchQueue.random.async {
+                value += 1
+            }
+        }
+        AssertEqualEventually(value.get(), numberOfStressTestCycles)
+
+        // Count down to zero
+        for _ in 0..<numberOfStressTestCycles {
+            DispatchQueue.random.async {
+                value -= 1
+            }
+        }
+        AssertEqualEventually(value.get(), 0)
+
+        // Check for memory leaks
+        weak var weakValue = value
+        value = nil
+
+        AssertNilEventually(weakValue)
```

Comment:
Why "eventually"? Shouldn't it be nil immediately?

Reply:
I noticed that for the original `DispatchQueue`-based solution the denin wasn't often happening immediately but (probably) in the next run loop. I didn't want to be that strict with tests here to allow alternative implementations. What's important is that it _eventually_ gets released :)

Reply:
If it's being released immediately in the new implementation, I'd say we can assume that's the expected behavior from now on. Something not being deallocated unexpectedly might lead to issues

Reply:
I just checked and even the current implementation is not released immediately 100% of the time so I have to keep it as-is. Not a big deal really.